### ISSUE TITLE
EgoDex Delta Action statistic result broken

### ISSUE THREAD

**Issue by csjiyw** [physical-superintelligence-lab/Psi0#33]
# EgoDex Delta Action Statistics Issue — Deep Analysis

> Generated: 2026-04-23  
> Purpose: Document all inconsistencies in the EgoDex delta action computation pipeline to inform next-step decisions

---

## I. Problem Summary

Psi0 uses `--data.use-delta-actions` during pretraining to convert absolute actions to deltas (frame-to-frame differences). However, the **statistics computation pipeline** for `egodex_stat_all.json` has **broken** through code evolution. Three interconnected issues exist:

| # | Issue | Severity | Status |
|---|-------|----------|--------|
| 1 | `--use_delta_actions` functionality removed from `calc_stat.py` (regression) | 🔴 High | Unfixed |
| 2 | Old stats use `camera[0]`, new training uses `camera[index]` (reference frame mismatch) | 🟡 Medium | Never consistent |
| 3 | `egodex_stat_delta.json` uses `quantile_01`/`quantile_99` instead of `q01`/`q99` (key mismatch) | 🟡 Medium | Format inconsistency |

---

### 2.1 Current Psi0 Training Pipeline (Runtime)

```
HDF5 (actions_48d, absolute actions)
    │
    ▼
egodex_dataset.py :: get_item()
    │  1. Read actions_48d[chunk_size+1 frames]
    │  2. cam_ext = camera[index]     ← Randomly sampled frame's camera
    │  3. convert_to_delta_actions(actions, cam_ext)
    │     └→ Convert to camera frame, then compute frame-to-frame differences
    ▼
delta_actions (chunk_size, 48)
    │
    ▼
ActionStateTransform :: normalize()
    │  stat_path = assets/stats/egodex_stat_all.json
    │  action_norm_type = bounds_q99
    │  action_min = stats["egodex"]["q01"]
    │  action_max = stats["egodex"]["q99"]
    │  normalized = (x - min) / (max - min) * 2 - 1
    ▼
normalized_delta_actions → into model
```

**Key config** (`scripts/train/psi0/pretrain-egodex-psi0-fast.sh`):
```bash
--data.use-delta-actions
--data.transform.field.action_norm_type=bounds_q99
--data.transform.field.stat-action-key=egodex
--data.transform.field.stat-path=assets/stats/egodex_stat_all.json
```

### 2.2 Current calc_stat.py (Broken)

```
HDF5 (actions_48d, absolute actions)
    │
    ▼
calc_stat.py (current version = commit 0d68a7f)
    │  ❌ No --use_delta_actions argument
    │  ❌ No convert_to_delta_actions()
    │  ❌ Only computes min, max (no q01/q99/mean/std)
    ▼
{"egodex": {"min": [...], "max": [...]}}    ← Absolute action statistics
```

### 2.3 Old calc_stat.py (commit 1e150bc)

```
HDF5 (actions_48d, absolute actions)
    │
    ▼
calc_stat.py (old version) --use_delta_actions
    │  cam_ext = camera[0]               ← Fixed: always frame 0
    │  convert_to_delta_actions(actions, cam_ext)
    │  Computes min, max, q01, q99, mean, std
    ▼
{"egodex": {"min": [...], "max": [...], "q01": [...], "q99": [...], "mean": [...], "std": [...]}}
```

---

## III. Git History Trace

```
cc6fc71 include hrdt preprocessing code
        ├── calc_stat.py first introduced, with full --use_delta_actions support
        │
        │ (10 intermediate commits, calc_stat.py unchanged)
        │
1e150bc add task_description_dict
        ├── calc_stat.py still has --use_delta_actions, upsample_rate,
        │   convert_to_delta_actions(), full stats (min/max/q01/q99/mean/std)
        │
        │ (10 more intermediate commits, calc_stat.py unchanged)
        │
0d68a7f pre-release SIMPLE for Psi-0
        └── ⚠️ Single commit stripped ALL delta-related code:
            - Removed --use_delta_actions argument
            - Removed --upsample_rate argument
            - Removed convert_to_delta_actions() and all helper functions
            - Removed q01, q99, mean, std computation
            - Only min, max remain

(Through HEAD = 6e21253)
        └── Never fixed ❌
```

**Conclusion**: This is a **regression bug** introduced during the "pre-release SIMPLE" cleanup commit. The author never fixed it.

---

### 4.1 `assets/stats/egodex_stat_all.json` (Currently used by training)

```json
{
  "egodex": {
    "min": [-0.467, ...],       // 48-dim
    "max": [0.411, ...],
    "q01": [-0.029, ...],       // Near zero → delta characteristic
    "q99": [0.030, ...],
    "mean": [4.85e-05, ...],    // Near zero → delta characteristic
    "std": [0.008, ...]
  }
}
```

- **Source**: Generated by **old** `calc_stat.py --use_delta_actions`
- **Verified**: `mean ≈ 0`, narrow `q01/q99` range (~±0.03) — classic delta signature.
  Absolute action means cannot be near zero (wrist positions have fixed offsets)
- **Camera reference**: `camera[0]` (first frame of each episode)
- **Key format**: `q01`/`q99` ✅ Matches Psi0 `ActionStateTransform`
- **Status**: **In active use**, training depends on this file

### 4.2 `src/h_rdt/datasets/pretrain/egodex_stat.json`

```json
{
  "egodex": {
    "min": [-0.986, ...],     // Wider range → absolute actions
    "max": [1.223, ...]
  }
}
```

- **Source**: Generated by **current** `calc_stat.py` (min/max only)
- **Camera reference**: N/A (no delta conversion, pure absolute)
- **Key format**: Only `min`/`max` ❌ No q01/q99, unusable for `bounds_q99` normalization
- **Purpose**: H-RDT pipeline (not Psi0)

### 4.3 `src/h_rdt/datasets/pretrain/egodex_stat_delta.json`

```json
{
  "egodex": {
    "min": [-0.415, ...],
    "max": [0.382, ...],
    "quantile_01": [...],      // ⚠️ Key is quantile_01, not q01
    "quantile_99": [...],      // ⚠️ Key is quantile_99, not q99
    "mean": [0.0001, ...],
    "std": [0.016, ...]
  }
}
```

- **Source**: H-RDT pipeline independent generation
- **Key format**: `quantile_01`/`quantile_99` ❌ Does NOT match Psi0's expected `q01`/`q99`
- **Purpose**: H-RDT training

### 4.4 `src/h_rdt/datasets/pretrain/egodex_stat_abs.json`

```json
{
  "egodex": {
    "min": [-0.726, ...],
    "max": [1.194, ...],
    "quantile_01": [...],
    "quantile_99": [...],
    "mean": [-0.141, ...],     // Clearly non-zero → absolute actions
    "std": [0.124, ...]
  }
}
```

- **Source**: H-RDT pipeline absolute action statistics
- **Purpose**: Reference

### 4.5 File Comparison Summary

| File | Type | Fields | Key format | Camera ref | Usable for Psi0 training |
|------|------|--------|------------|------------|-------------------------|
| `assets/stats/egodex_stat_all.json` | delta | All 6 | `q01`/`q99` ✅ | `camera[0]` | ✅ **Currently in use** |
| `h_rdt/.../egodex_stat.json` | absolute | min/max only | N/A | N/A | ❌ |
| `h_rdt/.../egodex_stat_delta.json` | delta | All 6 | `quantile_01` ❌ | ? | ❌ Key mismatch |
| `h_rdt/.../egodex_stat_abs.json` | absolute | All 6 | `quantile_01` ❌ | N/A | ❌ |

---

### Old stats: `camera[0]` (fixed per episode)

```
episode: [frame_0, frame_1, frame_2, ..., frame_N]
camera:  [cam_0,     cam_1,     cam_2,     ..., cam_N]

Stats computation: All frames converted to cam_0 coordinate system, then differenced
delta[t] = action_cam0[t+1] - action_cam0[t]
```

### New training: `camera[index]` (per random sample)

```
episode: [frame_0, ..., frame_N, ..., frame_M, ...]
                               ↑         ↑
                             index     index+1

Training: Uses cam_N as reference, only converts chunk_size+1 frames
delta[t] = action_camN[t+1] - action_camN[t]
```

### How significant is the difference?

EgoDex data uses a **head-mounted camera** capturing daily activities (cooking, typing, etc.). Within an episode:

- **Short time window (~1 sec, chunk_size=1, upsample_rate=3 → ~4 frames @30fps)**:
  Camera motion is minimal, `cam_N` and `cam_0` difference is negligible
- **Long time window (full episode 30-120 sec)**:
  Camera may shift significantly, `cam_0` and `cam_N` could differ by centimeters or more

**Key observation**:
- Stats use `camera[0]` for **all frames in the episode** — same global reference
- Training uses `camera[index]` for **each chunk** — each chunk has its own local reference

This means:
1. Training deltas are in a **local reference frame** — camera self-motion partially cancels out
2. Old stats deltas are in a **global reference frame** (`camera[0]`) — include camera drift over the whole episode

**Practical impact**: For `bounds_q99` normalization, we only care about the **distribution range** (q01/q99 values). Since training samples short chunks (1-4 frames), the local delta magnitude is in the same order as global delta magnitude. While values are not identical, using them as normalization bounds is an **acceptable approximation**.

---

## VIII. Open Questions

1. **Is full EgoDex stats recomputation needed?** If data hasn't changed, existing `egodex_stat_all.json` remains valid
2. **Should stats computation also incorporate `upsample_rate`?** Old version has it (default=6), training config uses `upsample_rate=3` — should these match?
3. **Should `h_rdt/.../egodex_stat_delta.json` keys be renamed?** From `quantile_01` to `q01` for Psi0 compatibility
4. **Should training script comments be updated?** Documenting stats computation method and limitations

---

## IX. Files Involved

| File | Needs modification |
|------|--------------------|
| `src/h_rdt/datasets/pretrain/calc_stat.py` | ✅ Primary fix target |
| `src/psi/data/egodex/egodex_dataset.py` | ❌ Training logic is correct, leave alone |
| `src/psi/config/transform.py` | ❌ Normalization logic is correct, leave alone |
| `scripts/train/psi0/pretrain-egodex-psi0-fast.sh` | ⚠️ May need comment updates |
| `assets/stats/egodex_stat_all.json` | ⚠️ Can continue using or regenerate |

**Comment by songlin** [comment]
Hi @csjiyw, it is amazing that you found the problem with such a detailed analysis. I found the history version of `src/h_rdt/datasets/pretrain/calc_stat.py` back. Everything related to the EgoDex delta action statistics should be consistent now.
